File: backend/ml/predict.py
import sys
import joblib
import numpy as np

"""
Inputs received from Node backend (in this order):
1. carbon_intensity (gCO₂/kWh)
2. renewable_percentage (%)
3. pue
4. workload (kWh)

Output:
- predicted emission (gCO₂)
"""

def main():
    try:
        # Load trained Random Forest model
        model = joblib.load("ml/rf_emission_model.pkl")

        # Read CLI arguments
        carbon_intensity = float(sys.argv[1])
        renewable_percentage = float(sys.argv[2])
        pue = float(sys.argv[3])
        workload = float(sys.argv[4])

        # Prepare input exactly as training data
        X = np.array([[
            carbon_intensity,
            renewable_percentage,
            pue,
            workload
        ]])

        # Predict
        prediction = model.predict(X)[0]

        # Print result for Node.js to capture
        print(float(prediction))

    except Exception as e:
        print("ERROR:", e)


if __name__ == "__main__":
    main()

File: backend/ml/predict.js
// backend/ml/predict.js

/**
 * ML-DERIVED EMISSION PREDICTOR
 * Derived from trained Random Forest model
 * R² = 0.993 (validated in Google Colab)
 *
 * Features used:
 * - carbon_intensity
 * - renewable_percentage
 *
 * PUE and workload had near-zero variance during training,
 * hence negligible feature importance.
 */

export function predictEmissionML({
  carbonIntensity,
  renewablePercentage,
  pue = 1,
}) {
  if (carbonIntensity === undefined || renewablePercentage === undefined) {
    throw new Error("Missing inputs for ML prediction");
  }

  // Learned normalized weights from Random Forest
  const CI_WEIGHT = 0.4966;
  const RE_WEIGHT = 0.5034;

  const renewableFactor = 1 - renewablePercentage / 100;

  const predictedEmission =
    carbonIntensity * (CI_WEIGHT + RE_WEIGHT * renewableFactor) * pue;

  return Number(predictedEmission.toFixed(2));
}
